src/pubsub.c

```diff
@@ -181,15 +300,55 @@ int pubsubUnsubscribeChannel(client *c, robj *channel, int notify) {
             /* Free the list and associated hash entry at all if this was
              * the latest client, so that it will be possible to abuse
              * Redis PUBSUB creating millions of channels. */
-            dictDelete(server.pubsub_channels,channel);
+            dictDelete(meta.serverPubSubChannels(), channel);
+            /* As this channel isn't subscribed by anyone, it's safe
+             * to remove the channel from the slot. */
+            if (server.cluster_enabled & meta.local) {
+                slotToChannelDel(channel->ptr);
+            }
         }
     }
     /* Notify the client */
-    if (notify) addReplyPubsubUnsubscribed(c,channel);
+    if (notify) {
+        addReplyPubsubUnsubscribed(c,channel,meta);
+    }
     decrRefCount(channel); /* it is finally safe to release it */
     return retval;
 }
 
+void pubsuLocalbUnsubscribeAllClients(robj *channel, int notify) {
+    int retval;
+    dictEntry *de = dictFind(server.pubsublocal_channels,channel);
+    serverAssertWithInfo(NULL,channel,de != NULL);
+    list *clients = dictGetVal(de);
+    if (listLength(clients) > 0) {
+        /* For each client subscribed to the channel, unsubscribe it. */
+        listIter li;
+        listNode *ln;
+        listRewind(clients, &li);
+        while ((ln = listNext(&li)) != NULL) {
+            client *c = listNodeValue(ln);
+            retval = dictDelete(c->pubsublocal_channels,channel);
+            serverAssertWithInfo(c,channel,retval == DICT_OK);
+            if (notify) {
+                addReplyPubsubUnsubscribed(c,channel,pubSubLocalMeta);
```

Comment:
I'm not sure clients know what to do about this. I would have expected we disconnect them and they can reconnect to the right client.

Reply:
Shouldn't we send at least a notification that the client is getting disconnected from this channel ?

Reply:
If they're getting disconnected I'm not sure it matters.